### DISCUSSION TITLE
expose state.value to other components/apps

### DISCUSSION THREAD

**Discussion by maninkari** [statelyai/xstate#3150] [Q&A]
Hi, 
I have a widget that connects to crypto wallets and uses x-state and a stateMachine for moving between the different states of the widget like: loading, authenticating, displaying, failure, etc. Now, I've been asked to expose the state of the widget (the stateMachine really) so that apps can react to it. What would be the easiest way of achieving this? I wouldn't like anyone to change the stateMachine from outside the widget. 

I was thinking of: 
- triggering an event from the widget every time state.value changes
- passing a callback function to the widget and calling it every time state.value changes
- use useService + context (still not very clear how it works)

Ideally I would like to have a hook 'useWidgetState' but I am not sure how to do this. Any feedback welcome.
Thank you!

**Comment by davidkpiano** [comment]
If you have access to the interpreted machine, you can listen for state changes:

```js
service.subscribe(state => {
  console.log(state.value);
});
```

This can be put on a `useEffect` or perhaps easier, via `useSelector`:

```js
const stateValue = useSelector(service, state => state.value);
```

Would this work for your use-case?

**Reply by maninkari** [reply]
Hi @davidkpiano, thank you for your reply. 

The widget is a react component published as an nm package and it should be a black box for the App using it. Therefore, I don't think how the App could get access to the state machine. 

In any case, I probably don't want to give the App access to the state machine inside the widget anyway, I only would like the App/other components to be able to react to the widget's state.value changes. 

I understand that for your suggestion a service should be instantiated on the App or in other component using the widget's state machine. Not sure how that could be achieved. Thank you!

<img width="406" alt="widget state machine" src="https://user-images.githubusercontent.com/3520378/158198169-7ab93d6c-ad19-4628-a695-61c29196f9a1.png">

**Reply by Andarist** [reply]
1. You can have like `MyStateProvider` that would wrap your app and your widget, making it possible to consume the state from just any place in your app. This would potentially require people to use both the provider and the widget component to make the whole thing work.
2. You can expose a callback on your widget and let people manage their own pub/sub mechanism

Either way, you probably want to only expose to the outside world a subset of what your machine knows but that can be achieved quite easily by just mapping the `machine` to whatever data you want.

**Reply by maninkari** [reply]
Thank you @Andarist! Seems like the easiest would be to pass a callback function.
